### PR TITLE
fix: Self-join optimization doesn't needlessly invalidate caching

### PR DESCRIPTION
**Pull request by TrevorBergeron**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: bigframes/core/rewrite.py [hunk 1]
```diff
@@ -311,3 +326,25 @@ def get_node_column_ids(node: nodes.BigFrameNode) -> Tuple[str, ...]:
     import bigframes.core
 
     return tuple(bigframes.core.ArrayValue(node).column_ids)
+
+
+def common_subtree(
+    l_tree: nodes.BigFrameNode, r_tree: nodes.BigFrameNode
+) -> Optional[nodes.BigFrameNode]:
+    """Find common subtree between join subtrees"""
```
Comment (chelsea-lin) [inline bigframes/core/rewrite.py:334]: To find common substree between two trees, the substree should have same node type and as well all children nodes should be same as well. IIUC, the below code can return a substree when its root node are same while its child nodes can be different. Another question here, is the `REWRITABLE_NODE_TYPES` unary nodes only? Can be binary node also?

File: bigframes/core/rewrite.py [hunk 2]
```diff
@@ -311,3 +326,25 @@ def get_node_column_ids(node: nodes.BigFrameNode) -> Tuple[str, ...]:
...
+        l_node = l_node.child
+    l_nodes.add(l_node)
+
+    r_node = r_tree
+    while isinstance(r_node, REWRITABLE_NODE_TYPES):
+        if r_node in l_nodes:
+            return r_node
+        r_node = r_node.child
+
+    if r_node in l_nodes:
```
Comment (chelsea-lin) [inline bigframes/core/rewrite.py:348]: With the above codes, this condition will never be triggered? Can you double check the test coverage?

**Review by chelsea-lin — CHANGES_REQUESTED** [review]

File: bigframes/core/rewrite.py (hunk 1 above)
Comment (TrevorBergeron) [reply bigframes/core/rewrite.py:334]: equality check is deep by default with dataclasses
try this for example. this returns `True`
```import dataclasses

@dataclasses.dataclass(frozen=True)
class Bar:
  a: int
  b: int

@dataclasses.dataclass(frozen=True)
class Foo:
  a: Bar


Foo(Bar(1,2)) == Foo(Bar(1,2))
```

Also, `REWRITABLE_NODE_TYPES` are the set of nodes that can be rewritten from a join into a projection.

File: bigframes/core/rewrite.py (hunk 2 above)
Comment (TrevorBergeron) [reply bigframes/core/rewrite.py:348]: the last line in the loop is `r_node = r_node.child`. this node isn't checked on the last iteration, so must check one last time. Not the cleanest code, but this line is needed. Probably a cleaner way of organizing this

Comment (chelsea-lin) [reply bigframes/core/rewrite.py:348]: Got it! Thanks!

File: bigframes/core/rewrite.py [hunk 3]
```diff
@@ -196,28 +203,36 @@ def expand(self) -> nodes.BigFrameNode:
...
-
 def maybe_rewrite_join(join_node: nodes.JoinNode) -> nodes.BigFrameNode:
-    left_side = SquashedSelect.from_node(join_node.left_child)
-    right_side = SquashedSelect.from_node(join_node.right_child)
+    rewrite_common_node = common_subtree(join_node.left_child, join_node.right_child)
+    if rewrite_common_node is None:
+        return join_node
+    left_side = SquashedSelect.from_node(join_node.left_child, rewrite_common_node)
+    right_side = SquashedSelect.from_node(join_node.right_child, rewrite_common_node)
     if left_side.can_join(right_side, join_node.join):
```
Comment (chelsea-lin) [inline bigframes/core/rewrite.py:212]: maybe rename it as `can_merge`, I thought it was checking if two trees are joinable or not.

**Review by chelsea-lin — APPROVED** [review]
LGTM but left comments to rename or add more comments to describe the optimization strategy.

File: bigframes/core/rewrite.py [hunk 4]
```diff
@@ -116,14 +123,14 @@ def can_join(
...
+    def merge(
         self,
         right: SquashedSelect,
         join_type: join_defs.JoinType,
         mappings: Tuple[join_defs.JoinColumnMapping, ...],
-    ) -> Optional[SquashedSelect]:
+    ) -> SquashedSelect:
         if self.root != right.root:
-            return None
+            raise ValueError("Cannot merge expressions with different roots")
```
[the remark this answers is not in the document]
Comment (TrevorBergeron) [reply bigframes/core/rewrite.py:133]: So far we don't do this for other similar errors - couldn't find `InternalError` anywhere in the code base. Maybe we should automatically wrap all exceptions with feedback link? Seems out of scope

File: bigframes/core/rewrite.py (hunk 3 above)
Comment (TrevorBergeron) [reply bigframes/core/rewrite.py:212]: done

File: bigframes/core/rewrite.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chelsea-lin) [reply bigframes/core/rewrite.py:134]: Could be `SystemError` in the python exceptions?

**Pull request merged**